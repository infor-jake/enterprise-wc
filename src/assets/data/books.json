[
  {
    "book": 101,
    "description": "101",
    "ledger": "",
    "bookCurrency": "usd",
    "transactionCurrency": "Book",
    "postHistory": false,
    "active": "Yes",
    "inStock": true,
    "convention": "Full Month",
    "methodSwitch": "Yes",
    "trackDeprecationHistory": "Yes",
    "useForEmployee": "No",
    "icon": "icon-edit",
    "image": "https://placekitten.com/200/100",
    "category": "Demo Category",
    "count": "3",
    "deprecationHistory": "Asset Label",
    "publishDate": "2021-04-23T18:25:43.511Z",
    "price": "12.99",
    "location": "United States",
    "color": "#165eab",
    "children":{
      "prop1": "row1_childprop1",
      "prop2": "row1_childprop2",
      "prop3": {
        "info": "row1_childprop3_info"
      }
    }
  },
  {
    "book": 102,
    "description": "102",
    "ledger": "CORE",
    "bookCurrency": "eur",
    "transactionCurrency": "Book",
    "postHistory": false,
    "active": "Yes",
    "inStock": true,
    "convention": "Full Month",
    "methodSwitch": "No",
    "trackDeprecationHistory": "No",
    "useForEmployee": "No",
    "icon": "icon-closed-folder",
    "category": "Demo Category",
    "count": "4.6",
    "deprecationHistory": "Asset Label",
    "publishDate": "2021-02-23T18:25:43.511Z",
    "price": "13.99",
    "location": "<img src='a' onerror='alert(0)'/>Canada",
    "color": "#35d783"
  },
  {
    "book": 103,
    "description": "103",
    "ledger": "CORE",
    "bookCurrency": "eur",
    "transactionCurrency": "Book",
    "postHistory": false,
    "active": "Yes",
    "inStock": false,
    "convention": "Half Month",
    "methodSwitch": "Yes",
    "trackDeprecationHistory": "No",
    "useForEmployee": "Yes",
    "icon": null,
    "image": "https://placekitten.com/400/300",
    "category": null,
    "deprecationHistory": "Asset Label",
    "publishDate": "2021-12-23T18:25:43.511Z",
    "location": "United States",
    "color": "orange",
    "price": "1.21"
  },
  {
    "book": 104,
    "description": "104",
    "ledger": "CORE",
    "bookCurrency": "usd",
    "transactionCurrency": "Book",
    "postHistory": false,
    "active": "No",
    "inStock": false,
    "convention": "2 Weeks",
    "methodSwitch": "No",
    "trackDeprecationHistory": "No",
    "useForEmployee": "No",
    "icon": "closed-folder",
    "category": "Demo Category",
    "count": "85",
    "deprecationHistory": "Asset Label",
    "location": "Canada",
    "color": "#f50c14",
    "price": "1.21"
  },
  {
    "book": 105,
    "description": "105",
    "ledger": "CORE",
    "bookCurrency": "usd",
    "transactionCurrency": "Book",
    "postHistory": false,
    "active": "Yes",
    "inStock": true,
    "convention": "Full Month",
    "methodSwitch": "Yes",
    "trackDeprecationHistory": "No",
    "useForEmployee": "Yes",
    "icon": "headphones",
    "category": "Demo Category",
    "count": 73,
    "deprecationHistory": "Asset Label",
    "publishDate": "2021-12-12T18:25:43.511Z",
    "price": "15.99",
    "location": "United States",
    "color": null
  },
  {
    "book": 106,
    "description": "106",
    "ledger": "CORE",
    "bookCurrency": "yen",
    "transactionCurrency": "Book",
    "postHistory": true,
    "active": "Yes",
    "inStock": false,
    "convention": "Full Month",
    "methodSwitch": "No",
    "trackDeprecationHistory": "No",
    "useForEmployee": "No",
    "icon": null,
    "category": null,
    "deprecationHistory": "Asset Label",
    "price": "1.21"
  },
  {
    "book": 107,
    "description": "107",
    "ledger": "CORE",
    "bookCurrency": "usd",
    "transactionCurrency": "Book",
    "postHistory": true,
    "active": "Yes",
    "inStock": false,
    "convention": "Full Month",
    "methodSwitch": "No",
    "trackDeprecationHistory": "No",
    "useForEmployee": "No",
    "icon": "icon-edit",
    "image": "https://placekitten.com/200/100",
    "category": "Demo Category",
    "count": 6,
    "deprecationHistory": "Asset Label",
    "location": "Canada",
    "color": "#8d1cf3",
    "price": "200.10"
  },
  {
    "book": 108,
    "description": "108",
    "ledger": "",
    "bookCurrency": "usd",
    "transactionCurrency": "Book",
    "postHistory": false,
    "active": "Yes",
    "inStock": true,
    "convention": "Full Month",
    "methodSwitch": "No",
    "trackDeprecationHistory": "Yes",
    "useForEmployee": "Yes",
    "icon": "special-item",
    "category": "Demo Category",
    "count": 9,
    "deprecationHistory": "Asset Label",
    "publishDate": "2020-12-12T18:25:43.511Z"
  },
  {
    "book": 109,
    "description": "109",
    "ledger": null,
    "bookCurrency": "usd",
    "transactionCurrency": "Book",
    "postHistory": false,
    "active": "Yes",
    "inStock": false,
    "convention": "Full Month",
    "methodSwitch": "Yes",
    "trackDeprecationHistory": "No",
    "useForEmployee": "No",
    "icon": "icon-special-item",
    "category": "Demo Category",
    "count": 900,
    "deprecationHistory": "Asset Label",
    "publishDate": "2019-12-12T18:25:43.511Z",
    "location": "United States",
    "color": "#F00"
  }
]
