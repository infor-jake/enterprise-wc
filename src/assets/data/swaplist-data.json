[
  {
    "id": 1,
    "name": "Available",
    "items": [
      {
        "id": 1,
        "text": "Option A",
        "value": "option_a"
      },
      {
        "id": 2,
        "text": "Option B",
        "value": "option_b"
      }
    ]
  },
  {
    "id": 2,
    "name": "Selected",
    "items": [
      {
        "id": 3,
        "text": "Option C",
        "value": "option_c"
      },
      {
        "id": 4,
        "text": "Option D",
        "value": "option_d"
      }
    ]
  },
  {
    "id": 3,
    "name": "Additional",
    "items": [
      {
        "id": 5,
        "text": "Option E",
        "value": "option_e"
      },
      {
        "id": 6,
        "text": "Option F",
        "value": "option_f"
      }
    ]
  }
]