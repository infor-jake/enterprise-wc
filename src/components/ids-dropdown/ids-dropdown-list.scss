/* Ids Dropdown Component */
@use '../../themes/mixins/ids-validation-mixin';

// Dropdown menu width
:host ids-popup {
  max-width: var(--ids-input-width-40);
}

:host([size='xs']) ids-popup {
  max-width: var(--ids-input-width-10);
}

:host([size='sm']) ids-popup {
  max-width: var(--ids-input-width-20);
}

:host([size='mm']) ids-popup {
  max-width: var(--ids-input-width-30);
}

:host([size='md']) ids-popup {
  max-width: var(--ids-input-width-40);
}

:host([size='lg']) ids-popup {
  max-width: var(--ids-input-width-50);
}

:host([size='full']) ids-popup {
  max-width: var(--ids-input-width-full);
}

:host ids-popup,
:host([size='full']) {
  width: 100%;
}
