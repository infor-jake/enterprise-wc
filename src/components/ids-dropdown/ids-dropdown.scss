/* Ids Dropdown Component */
@import '../../themes/mixins/ids-validation-mixin';

:host {
  display: block;

  .ids-dropdown {
    display: inline-block;
    max-width: 100%;
  }
}

ids-trigger-field[readonly-background]:not([disabled])::part(input) {
  cursor: var(--ids-cursor-pointer);
}

// Most Styles are in list-box, popup and trigger field
.ids-dropdown {
  display: block;
  position: relative;

  ids-trigger-field {
    &[size='full'] {
      display: block;
    }
  }

  &.full {
    display: block;

    ids-trigger-field,
    .ids-trigger-field {
      display: block;
    }
  }
}

// Icon Placement
.icon-container {
  color: var(--ids-dropdown-color-icon);
  margin: 0 0 0 var(--ids-dropdown-space-text);
  position: relative;
  top: 2px;
}

:host([mode='light']) {
  .typeahead.is-open ids-trigger-field::part(field-container) {
    background-color: var(--ids-dropdown-typeahead-color-background);
  }
}

:host([dir='rtl']) .icon-container {
  margin: 0 var(--ids-dropdown-space-text) 0 0;
}

::slotted(.sub-group) {
  border-top: 1px solid var(--ids-dropdown-separator-color-border);
  padding-top: var(--ids-dropdown-separator-padding);
}

// Dropdown Editor
.ids-dropdown.color-variant-borderless,
.ids-dropdown.color-variant-in-cell {
  ids-trigger-field::part(field-container) {
    margin: 0;
    border-radius: 0;
  }

  ids-trigger-field::part(label) {
    display: none;
  }
}

// Dropdown Editor Borderless
.ids-dropdown.color-variant-borderless {
  ids-trigger-field::part(container) {
    margin: 0;
  }

  ids-trigger-field::part(field-container) {
    box-shadow: none !important;
  }

  &.field-height-lg {
    ids-trigger-field::part(input) {
      padding-inline: 16px;
      height: 50px;
    }
  }

  &.field-height-md {
    ids-trigger-field::part(input) {
      padding-inline: 12px;
      height: 40px;
    }
  }

  &.field-height-sm {
    ids-trigger-field::part(input) {
      padding-inline: 8px;
      height: 35px;
    }
  }

  &.field-height-xs {
    ids-trigger-field::part(input) {
      padding-inline: 4px;
      height: 30px;
    }
  }
}

// Dropdown Editor In Cell
.ids-dropdown.color-variant-in-cell {
  ids-trigger-field::part(container) {
    margin: 1px 0;
  }
}

// Dropdown menu width
:host ids-dropdown-list {
  max-width: var(--ids-input-width-40);
}

:host([size='xs']) ids-dropdown-list {
  max-width: var(--ids-input-width-10);
}

:host([size='sm']) ids-dropdown-list {
  max-width: var(--ids-input-width-20);
}

:host([size='mm']) ids-dropdown-list {
  max-width: var(--ids-input-width-30);
}

:host([size='md']) ids-dropdown-list {
  max-width: var(--ids-input-width-40);
}

:host([size='lg']) ids-dropdown-list {
  max-width: var(--ids-input-width-50);
}

:host([size='full']) ids-dropdown-list {
  max-width: var(--ids-input-width-full);
}

:host ids-dropdown-list,
:host([size='full']) {
  width: var(--ids-input-width-full);
}

// Change state for external dropdown list
:host([list]:not([disabled])) {
  cursor: var(--ids-cursor-pointer);
}
