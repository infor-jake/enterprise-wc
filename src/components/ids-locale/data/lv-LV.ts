// Info Comes from http://www.unicode.org/Public/cldr/25/
const locale = {
  // layout/language
  language: 'lv',
  englishName: 'Latvian (Latvia)',
  nativeName: 'latviešu (Latvija)',
  // layout/orientation/@characters
  direction: 'left-to-right',
  // ca-gregorian
  calendars: [{
    name: 'gregorian',
    // ca-gregorian/main/dates/calendars/gregorian/dateFormats/
    dateFormat: {
      separator: '.', // Infered
      timeSeparator: ':',
      short: 'dd.MM.yyyy.', // use four digit year
      medium: 'MMM yyyy',
      long: 'd. MMMM yyyy',
      full: 'EEEE, y. gada d. MMMM',
      month: 'd. MMMM',
      year: 'MMMM yyyy',
      dayOfWeek: 'EEE d',
      timestamp: 'HH:mm:ss',
      timestampMillis: 'HH:mm:ss.SSS',
      hour: 'HH:mm',
      datetime: 'dd.MM.yyyy HH:mm',
      datetimeMillis: 'dd.MM.yyyy HH:mm:ss.SSS',
      timezone: 'dd.MM.yyyy HH:mm zz',
      timezoneLong: 'dd.MM.yyyy HH:mm zzzz'
    }, // Infered short + short gregorian/dateTimeFormats
    // ca-gregorian/main/dates/calendars/gregorian/days/format/short or abbreviated (2 digit)
    days: {
      wide: ['svētdiena', 'pirmdiena', 'otrdiena', 'trešdiena', 'ceturtdiena', 'piektdiena', 'sestdiena'],
      abbreviated: ['Sv', 'Pr', 'Ot', 'Tr', 'Ce', 'Pk', 'Se'],
      narrow: ['S', 'P', 'O', 'T', 'C', 'P', 'S']
    },
    // ca-gregorian/main/dates/calendars/gregorian/months/format/wide
    months: {
      wide: ['Janvāris', 'Februāris', 'Marts', 'Aprīlis', 'Maijs', 'Jūnijs', 'Jūlijs', 'Augusts', 'Septembris', 'Oktobris', 'Novembris', 'Decembris'],
      abbreviated: ['Janv.', 'Febr.', 'Marts', 'Apr.', 'Maijs', 'Jūn.', 'Jūl.', 'Aug.', 'Sept.', 'Okt.', 'Nov.', 'Dec.']
    },
    // ca-gregorian/main/dates/calendars/gregorian/timeFormats/short
    timeFormat: 'HH:mm',
    // ca-gregorian/main/dates/calendars/gregorian/dayPeriods/wide
    dayPeriods: ['priekšpusdienā', 'pēcpusdienā'],
    firstDayofWeek: 1 // Starts on Monday
  }],
  // numbers/currencyFormats-numberSystem-latn/standard
  currencySign: 'Ls',
  currencyFormat: '¤###',
  // numbers/symbols-numberSystem-latn
  numbers: {
    percentSign: '%',
    percentFormat: '### %',
    percentSuffix: ' %',
    percentPrefix: '',
    minusSign: '-',
    decimal: ',',
    group: ' ',
    groupSizes: [3, 3]
  }
};

export { locale };
