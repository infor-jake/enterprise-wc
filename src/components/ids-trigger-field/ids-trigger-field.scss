@import '../ids-input/ids-input';

// Ability to center label
:host([part='trigger-field']) {
  justify-content: flex-start;
}

.ids-trigger-field {
  justify-content: inherit;

  &.compact .field-container {
    height: var( --ids-input-height-10);
  }
}

.ids-trigger-field-slot-trigger-start,
::slotted(*[slot='trigger-start']:first-of-type:not([inline]):not([compact]):not([field-height='xs'])) {
  margin-inline-start: var(--ids-trigger-field-button-margin);
}

.ids-trigger-field-slot-trigger-end,
::slotted(*[slot='trigger-end']:last-of-type:not([inline]):not([compact]):not([field-height='xs'])) {
  margin-inline-end: var(--ids-trigger-field-button-margin);
}

.ids-input.color-variant-alternate-formatter {
  &:focus-within {
    .field-container {
      box-shadow: var(--ids-trigger-field-shadow-focus);
    }

    .ids-input-field {
      box-shadow: none;
    }
  }
}
