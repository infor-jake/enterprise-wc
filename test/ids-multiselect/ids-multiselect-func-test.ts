/**
 * @jest-environment jsdom
 */
import '../helpers/resize-observer-mock';
import waitForTimeout from '../helpers/wait-for-timeout';
import wait from '../helpers/wait';
import processAnimFrame from '../helpers/process-anim-frame';

import IdsMultiselect from '../../src/components/ids-multiselect/ids-multiselect';
import '../../src/components/ids-dropdown/ids-dropdown';
import '../../src/components/ids-dropdown/ids-dropdown-list';
import '../../src/components/ids-list-box/ids-list-box';
import '../../src/components/ids-list-box/ids-list-box-option';
import '../../src/components/ids-trigger-field/ids-trigger-field';
import states from '../../src/assets/data/states.json';

import IdsContainer from '../../src/components/ids-container/ids-container';
import { messages as deMessages } from '../../src/components/ids-locale/data/de-messages';
import IdsGlobal from '../../src/components/ids-global/ids-global';

describe('IdsMultiselect Component', () => {
  let multiselect: any;
  let container: any;

  const createFromTemplate = (innerHTML:any) => {
    multiselect?.remove();
    container?.remove();

    container = new IdsContainer();

    const template = document.createElement('template');
    template.innerHTML = innerHTML;
    multiselect = template.content.childNodes[0];

    container.appendChild(multiselect);
    document.body.appendChild(container);
    return multiselect;
  };

  beforeEach(async () => {
    multiselect = createFromTemplate(
      `<ids-multiselect id="multiselect-1" label="Normal Multiselect" dirty-tracker="true">
        <ids-list-box>
          <ids-list-box-option value="opt2" id="opt2" selected><ids-checkbox label="Option Two" tooltip="Additional Info on Option Two"></ids-checkbox></ids-list-box-option>
          <ids-list-box-option value="opt3" id="opt3"><ids-checkbox label="Option Three" tooltip="Additional Info on Option Three"></ids-checkbox></ids-list-box-option>
          <ids-list-box-option value="opt4" id="opt4"><ids-checkbox label="Option Four" tooltip="Additional Info on Option Four"></ids-checkbox></ids-list-box-option>
          <ids-list-box-option value="opt5" id="opt5"><ids-checkbox label="Option Five" tooltip="Additional Info on Option Five"></ids-checkbox></ids-list-box-option>
          <ids-list-box-option value="opt6" id="opt6"><ids-checkbox label="Option Six" tooltip="Additional Info on Option Six"></ids-checkbox></ids-list-box-option>
        </ids-list-box>
      </ids-multiselect>`
    );
  });

  afterEach(async () => {
    document.body.innerHTML = '';
  });

  it('renders empty multiselect with no errors', () => {
    const errors = jest.spyOn(global.console, 'error');
    multiselect.remove();
    const elem:any = new IdsMultiselect();
    document.body.appendChild(elem);
    elem.remove();
    expect(document.querySelectorAll('ids-multiselect').length).toEqual(0);
    expect(errors).not.toHaveBeenCalled();
  });

  it('renders correctly', () => {
    expect(multiselect.outerHTML).toMatchSnapshot();
  });

  it('renders with disabled', () => {
    multiselect = createFromTemplate(`<ids-multiselect id="multiselect-1" disabled="true" label="Test"></ids-multiselect>`);
    expect(multiselect.disabled).toBeTruthy();
  });

  it('renders with readonly', () => {
    multiselect = createFromTemplate(`<ids-multiselect id="multiselect-1" readonly="true" label="Test"></ids-multiselect>`);
    expect(multiselect.readonly).toBeTruthy();
  });

  it('can set readonly', () => {
    multiselect.readonly = true;
    expect(multiselect.readonly).toEqual(true);
    expect(multiselect.getAttribute('readonly')).toBeTruthy();

    multiselect.readonly = false;
    expect(multiselect.readonly).toEqual(false);
    expect(multiselect.getAttribute('readonly')).toBeFalsy();

    multiselect.setAttribute('readonly', 'true');
    expect(multiselect.readonly).toEqual(true);
    expect(multiselect.getAttribute('readonly')).toBeTruthy();

    multiselect.removeAttribute('readonly');
    expect(multiselect.readonly).toEqual(false);
    expect(multiselect.getAttribute('readonly')).toBeFalsy();
  });

  it('can set disabled', () => {
    multiselect.disabled = true;
    expect(multiselect.disabled).toEqual(true);
    expect(multiselect.getAttribute('disabled')).toBeTruthy();

    multiselect.disabled = false;
    expect(multiselect.disabled).toEqual(false);
    expect(multiselect.getAttribute('disabled')).toBeFalsy();

    multiselect.setAttribute('disabled', 'true');
    expect(multiselect.disabled).toEqual(true);
    expect(multiselect.getAttribute('disabled')).toBeTruthy();

    multiselect.removeAttribute('disabled');
    expect(multiselect.disabled).toEqual(false);
    expect(multiselect.getAttribute('disabled')).toBeFalsy();
  });

  it('renders with validation', () => {
    multiselect = createFromTemplate(`<ids-multiselect id="multiselect-1" validate="required" validation-events="blur change" label="Test"></ids-multiselect>`);
    expect(multiselect.validate).toEqual('required');
    expect(multiselect.validationEvents).toEqual('blur change');

    // Generate from the parent defaults
    multiselect = createFromTemplate(`<ids-multiselect id="multiselect-1" validate="required" label="Test"></ids-multiselect>`);
    multiselect.validationEvents = 'blur change';
    expect(multiselect.validate).toEqual('required');
    expect(multiselect.validationEvents).toEqual('blur change');

    // Default Case
    multiselect = createFromTemplate(`<ids-multiselect id="multiselect-1" validate="required" label="Test"></ids-multiselect>`);
    expect(multiselect.validate).toEqual('required');
    expect(multiselect.validationEvents).toEqual('change');
  });

  it('supports validation', async () => {
    multiselect = createFromTemplate(`<ids-multiselect id="multiselect-5" label="Dropdown with Icons" validate="true">
     </ids-multiselect>`);
    await waitForTimeout(() => expect(multiselect.shadowRoot.querySelector('ids-trigger-field')).toBeTruthy());

    multiselect.validate = 'required';
    multiselect.validationEvents = 'blur change';
    multiselect.triggerEvent('change', multiselect);
    expect(multiselect.getAttribute('validate')).toEqual('required');
  });

  it('can reset validation and validation-events', async () => {
    multiselect.validate = 'required';
    multiselect.validationEvents = 'blur change';
    multiselect.validate = null;
    multiselect.validationEvents = null;
    expect(multiselect.getAttribute('validate')).toBeFalsy();
    expect(multiselect.getAttribute('validation-events')).toBeFalsy();
  });

  it('handles setting disabled', () => {
    multiselect.disabled = true;
    expect(multiselect.getAttribute('disabled')).toEqual('true');
    expect(multiselect.getAttribute('readonly')).toBeFalsy();
    expect(multiselect.disabled).toEqual(true);
    expect(multiselect.input.disabled).toEqual(true);
  });

  it('handles setting readonly', () => {
    multiselect.readonly = true;
    expect(multiselect.getAttribute('readonly')).toEqual('true');
    expect(multiselect.readonly).toEqual(true);
    expect(multiselect.input.disabled).toEqual(false);
  });

  it('can change the label', () => {
    multiselect.label = 'Changed Label';
    expect(multiselect.label).toEqual('Changed Label');
  });

  it('should be able to reset dirty indicator', () => {
    multiselect.dirtyTracker = true;
    expect(multiselect.getAttribute('dirty-tracker')).toEqual('true');
    multiselect.dirtyTracker = false;
    expect(multiselect.getAttribute('dirty-tracker')).toBeFalsy();
  });

  it('should be able to set value', () => {
    multiselect.value = ['opt3'];
    expect(multiselect.value).toContain('opt3');
  });

  it('should ignore null / bad value', () => {
    multiselect.value = ['opt3'];
    expect(multiselect.value).toContain('opt3');

    multiselect.value = null;
    expect(multiselect.value).toContain('opt3');

    multiselect.value = ['optx'];
    expect(multiselect.value).toContain('opt3');
  });

  it('supports opening the list with open', async () => {
    expect(multiselect.popup.visible).toEqual(false);
    await multiselect.open();
    expect(multiselect.popup.visible).toEqual(true);
    multiselect.close();
    expect(multiselect.popup.visible).toEqual(false);
    multiselect.readonly = true;
    await multiselect.open();
    expect(multiselect.popup.visible).toEqual(false);
    multiselect.disabled = true;
    await multiselect.open();
    expect(multiselect.popup.visible).toEqual(false);
  });

  it('supports opening the list with toggle', async () => {
    expect(multiselect.popup.visible).toEqual(false);
    multiselect.toggle();
    expect(multiselect.popup.visible).toEqual(true);
    multiselect.toggle();
    expect(multiselect.popup.visible).toEqual(false);
  });

  it('supports closing the list with closing', async () => {
    expect(multiselect.popup.visible).toEqual(false);
    multiselect.open();
    expect(multiselect.popup.visible).toEqual(true);
    multiselect.close();
    expect(multiselect.popup.visible).toEqual(false);
    multiselect.open();
    expect(multiselect.popup.visible).toEqual(true);
    multiselect.close(true);
    expect(multiselect.popup.visible).toEqual(false);
  });

  it('can click outside an open list to close it', (done) => {
    const clickEvent = new MouseEvent('click', { bubbles: true });

    multiselect.dropdownList.onOutsideClick = jest.fn();
    multiselect.open();

    setTimeout(() => {
      // Click outside the Modal into the overlay area
      document.body.dispatchEvent(clickEvent);

      setTimeout(() => {
        expect(multiselect.dropdownList.onOutsideClick).toHaveBeenCalled();
        done();
      });
    }, 70);
  });

  it('supports async beforeShow', async () => {
    const getContents = () => new Promise((resolve) => {
      setTimeout(() => {
        resolve(states);
      }, 1);
    });

    expect(multiselect.querySelectorAll('ids-list-box-option').length).toEqual(5);
    expect(multiselect.beforeShow).toBeFalsy();
    multiselect.beforeShow = async function beforeShow() {
      return getContents();
    };
    expect(multiselect.beforeShow).toBeTruthy();
    await multiselect.open();
    expect(await multiselect.querySelectorAll('ids-list-box-option').length).toEqual(59);
  });

  it('ignores type ahead to open when no matches', async () => {
    multiselect.triggerEvent('keydownend', multiselect, { detail: { keys: 'xxxxx' } });

    await waitForTimeout(() => expect(multiselect.popup.visible).toEqual(false));
    expect(multiselect.popup.visible).toEqual(false);
  });

  it('ignores type ahead when readonly', async () => {
    multiselect.readonly = true;
    multiselect.triggerEvent('keydownend', multiselect, { detail: { keys: 'option thr' } });

    multiselect.disabled = true;
    multiselect.triggerEvent('keydownend', multiselect, { detail: { keys: 'option thr' } });

    await waitForTimeout(() => expect(multiselect.popup.visible).toEqual(false));
    expect(multiselect.popup.visible).toEqual(false);
  });

  it('supports clicking trigger to open', async () => {
    await waitForTimeout(() => expect(multiselect.trigger).toBeTruthy());

    multiselect.trigger.click();
    expect(multiselect.popup.visible).toEqual(true);
  });

  it('supports clicking input to open', async () => {
    await waitForTimeout(() => expect(multiselect.container).toBeTruthy());
    multiselect.input.shadowRoot.querySelector('.field-container').click();
    await waitForTimeout(() => expect(multiselect.popup.visible).toBeTruthy());
    expect(multiselect.popup.visible).toEqual(true);
  });

  it('supports clicking to select', async () => {
    expect(multiselect.value).toContain('opt2');
    multiselect.trigger.click();

    await wait(80);

    multiselect.querySelectorAll('ids-list-box-option')[4].click();

    await wait(80);
    expect(multiselect.value).toContain('opt6');
    expect(multiselect.value).toContain('opt2');
  });

  it('can changing language from the container', async () => {
    IdsGlobal.getLocale().loadedLanguages.set('de', deMessages);
    await IdsGlobal.getLocale().setLanguage('de');
    await processAnimFrame();
    expect(multiselect.getAttribute('aria-description')).toEqual('Drücken Sie zum Auswählen die Nach-unten-Taste');
  });

  it('opens on arrow down', () => {
    const event = new KeyboardEvent('keydown', { key: 'ArrowDown' });
    multiselect.dispatchEvent(event);

    expect(multiselect.popup.visible).toEqual(true);
  });

  it('ignores arrow down on open', () => {
    multiselect.open();
    const event = new KeyboardEvent('keydown', { key: 'ArrowDown' });
    multiselect.dispatchEvent(event);

    expect(multiselect.popup.visible).toEqual(true);
  });

  it('opens on arrow up', () => {
    const event = new KeyboardEvent('keydown', { key: 'ArrowUp' });
    multiselect.dispatchEvent(event);

    expect(multiselect.popup.visible).toEqual(true);
  });

  it('opens on enter', () => {
    const event = new KeyboardEvent('keydown', { key: 'Enter' });
    multiselect.dispatchEvent(event);

    expect(multiselect.popup.visible).toEqual(true);
  });

  it.skip('selects on space/enter when open', () => {
    multiselect.value = [];
    multiselect.open();
    multiselect.dropdownList.dispatchEvent(new KeyboardEvent('keydown', { key: 'ArrowDown' }));
    multiselect.dropdownList.dispatchEvent(new KeyboardEvent('keydown', { key: 'ArrowDown' }));
    multiselect.dropdownList.dispatchEvent(new KeyboardEvent('keydown', { key: 'ArrowDown' }));
    multiselect.dropdownList.dispatchEvent(new KeyboardEvent('keydown', { key: ' ' }));

    multiselect.close();
    expect(multiselect.value).toEqual(['opt5']);
  });

  it('should set/unset tags attribute', () => {
    multiselect.tags = true;
    expect(multiselect.tags).toBeTruthy();
    multiselect.tags = false;
    expect(multiselect.tags).toBeFalsy();
  });

  it('should set/unset max attribute', () => {
    multiselect.max = 5;
    expect(multiselect.max).toEqual(5);
    multiselect.max = null;
    expect(multiselect.max).toBeNaN();
  });

  it('tags work correctly', async () => {
    createFromTemplate(`<ids-multiselect id="multiselect-1" tags="true" label="Tags Multiselect" dirty-tracker="true">
      <ids-list-box>
        <ids-list-box-option value="opt1"><ids-checkbox label="Option One" class="justify-center"></ids-checkbox></ids-list-box-option>
        <ids-list-box-option value="opt2"><ids-checkbox label="Option Two" class="justify-center"></ids-checkbox></ids-list-box-option>
        <ids-list-box-option value="opt3" id="opt3"><ids-checkbox label="Option Three" class="justify-center"></ids-checkbox></ids-list-box-option>
        <ids-list-box-option value="opt4" id="opt4"><ids-checkbox label="Option Four" class="justify-center"></ids-checkbox></ids-list-box-option>
        <ids-list-box-option value="opt5" id="opt5"><ids-checkbox label="Option Five" class="justify-center"></ids-checkbox></ids-list-box-option>
        <ids-list-box-option value="opt6" id="opt6"><ids-checkbox label="Option Six" class="justify-center"></ids-checkbox></ids-list-box-option>
      </ids-list-box>
    </ids-multiselect>`);
    multiselect.value = ['opt1'];
    expect(multiselect.input.querySelectorAll('ids-tag').length).toEqual(1);

    multiselect.input.querySelector('ids-icon')?.dispatchEvent(new MouseEvent('click'));
    expect(multiselect.value).toEqual([]);

    multiselect.value = ['opt1'];
    multiselect.disabled = true;
    multiselect.value = ['opt1', 'opt2'];

    expect([...multiselect.input.querySelectorAll('ids-tag')].every((item) => item.hasAttribute('disabled'))).toBeTruthy();

    multiselect.disabled = false;
    multiselect.value = [];
    multiselect.open();

    multiselect.querySelector('ids-list-box-option')?.click();

    expect(multiselect.value).toEqual(['opt1']);
  });

  it('should handle overflowed text', async () => {
    multiselect = createFromTemplate(
      `<ids-multiselect label="Overflow text multiselect">
        <ids-list-box>
          <ids-list-box-option value="opt1" id="opt1" selected>Option One</ids-list-box-option>
          <ids-list-box-option value="opt2" id="opt2" selected>Option Two</ids-list-box-option>
          <ids-list-box-option value="opt3" id="opt3" selected>Option Three</ids-list-box-option>
        </ids-list-box>
      </ids-multiselect>`
    );
    const getText = () => multiselect.input.querySelector('ids-text');

    expect(multiselect.value).toEqual(['opt1', 'opt2', 'opt3']);
    expect(getText()?.textContent).toEqual('Option One, Option Two, Option Three');

    multiselect.value = ['opt1'];

    expect(getText()?.textContent).toEqual('Option One');

    multiselect.value = [];

    expect(getText()?.textContent).toBe('');
  });
});
