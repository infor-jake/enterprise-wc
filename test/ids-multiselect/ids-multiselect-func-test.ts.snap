// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`IdsMultiselect Component renders correctly 1`] = `
"<ids-multiselect id="multiselect-1" label="Normal Multiselect" dirty-tracker="true" role="combobox" aria-expanded="false" aria-autocomplete="list" aria-haspopup="listbox" aria-description="Press down arrow to select" aria-controls="ids-list-box-multiselect-1">
        <ids-list-box size="md" id="ids-list-box-multiselect-1" aria-label="Listbox" role="listbox">
      <ids-list-box-option class="multiselect-option" id="opt2" value="opt2" role="option" tabindex="-1">
        <ids-checkbox no-margin="" class="justify-center multiselect-checkbox" label="" checked="true"></ids-checkbox>
      </ids-list-box-option>
      <ids-list-box-option class="multiselect-option multiselect-border" id="opt3" value="opt3" role="option" tabindex="-1">
        <ids-checkbox no-margin="" class="justify-center multiselect-checkbox" label=""></ids-checkbox>
      </ids-list-box-option>
      <ids-list-box-option class="multiselect-option" id="opt4" value="opt4" role="option" tabindex="-1">
        <ids-checkbox no-margin="" class="justify-center multiselect-checkbox" label=""></ids-checkbox>
      </ids-list-box-option>
      <ids-list-box-option class="multiselect-option" id="opt5" value="opt5" role="option" tabindex="-1">
        <ids-checkbox no-margin="" class="justify-center multiselect-checkbox" label=""></ids-checkbox>
      </ids-list-box-option>
      <ids-list-box-option class="multiselect-option" id="opt6" value="opt6" role="option" tabindex="-1">
        <ids-checkbox no-margin="" class="justify-center multiselect-checkbox" label=""></ids-checkbox>
      </ids-list-box-option></ids-list-box>
      </ids-multiselect>"
`;
